types/aws-lambda/index.d.ts

```diff
@@ -576,6 +577,109 @@ export interface CodePipelineEvent {
     };
 }
 
+/**
+ * CodePipeline CloudWatch Events
+ * https://docs.aws.amazon.com/codepipeline/latest/userguide/detect-state-changes-cloudwatch-events.html
+ *
+ * The above CodePipelineEvent is when a lambda is invoked by a CodePipeline.
+ * These events are when you subsribe to CodePipeline events in CloudWatch.
+ *
+ * Their documentation says that detail.version is a string, but it is actually an integer
+ */
+export type CodePipelineState =
+    | 'STARTED'
+    | 'SUCCEEDED'
+    | 'RESUMED'
+    | 'FAILED'
+    | 'CANCELED'
+    | 'SUPERSEDED';
+
+export type CodePipelineStageState =
+    | 'STARTED'
+    | 'SUCCEEDED'
+    | 'RESUMED'
+    | 'FAILED'
+    | 'CANCELED';
+
+export type CodePipelineActionState =
+    | 'STARTED'
+    | 'SUCCEEDED'
+    | 'FAILED'
+    | 'CANCELED';
+
+export interface CodePipelinePiplelineEvent {
+    version: string;
+    id: string;
+    'detail-type': 'CodePipeline Pipeline Execution State Change';
+    source: 'aws.codepipeline';
+    account: string;
+    time: string;
+    region: string;
+    resources: string[];
+    detail: {
+        pipeline: string;
+        version: number;
+        state: CodePipelineState;
+        'execution-id': string;
+    };
+}
+
+export interface CodePipelineStageEvent {
+    version: string;
+    id: string;
+    'detail-type': 'CodePipeline Stage Execution State Change';
+    source: 'aws.codepipeline';
+    account: string;
+    time: string;
+    region: string;
+    resources: string[];
+    detail: {
+        pipeline: string;
+        version: number;
+        'execution-id': string;
+        stage: string;
+        state: CodePipelineStageState;
+    };
+}
+
+export type CodePipelineActionCategory =
+    | 'Approval'
+    | 'Build'
+    | 'Deploy'
+    | 'Invoke'
+    | 'Source'
+    | 'Test';
+
+export interface CodePipelineActionEvent {
+    version: string;
+    id: string;
+    'detail-type': 'CodePipeline Action Execution State Change';
+    source: 'aws.codepipeline';
+    account: string;
+    time: string;
+    region: string;
+    resources: string[];
+    detail: {
+        pipeline: string;
+        version: number;
+        'execution-id': string;
+        stage: string;
+        action: string;
+        state: CodePipelineActionState;
+        type: {
+            owner: 'AWS' | 'Custom' | 'ThirdParty';
+            category: CodePipelineActionCategory;
+            provider: 'CodeDeploy';
+            version: number;
+        };
+    };
+}
+
+export type CodePipelineCloudWatchEvent =
+    | CodePipelinePiplelineEvent
+    | CodePipelineStageEvent
+    | CodePipelineActionEvent;
```

Comment:
Since you could configure to receive only one of, or all of these events, this naming probably makes sense, but consider including `CloudWatch` in order to further distinguish from the (unfortunately named) `CodePipelineEvent`?

As mentioned by @jagregory, any `FooEvent` that doesn't have a result should for the user's convenience have a `export type FooHandler = Handler<FooEvent, void>`, declared at the bottom of this file (ones with results also should have a FooCallback<FooResult>`).

Here this should mean `CodePipelineCloudWatchHandler` as well as `CodePipelinePipelineHandler`, `CodePipelineStageHandler` and `CodePipelineActionHandler`, since it's reasonable to make a handler for all of them, switched, or separate handlers.

Reply:
Rename like this?

- `CodePipelinePipelineEvent` -> `CodePipelineCloudWatchPipelineEvent`
- `CodePipelineStageEvent` -> `CodePipelineCloudWatchStageEvent`
- `CodePipelineActionEvent` -> `CodePipelineCloudWatchActionEvent`

Reply:
Done those renames now anyway